File: app/src/main/java/com/example/coingryphon/MainActivity.java
package com.example.coingryphon;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.example.coingryphon.serial.GryphonProtocol;
import com.example.coingryphon.serial.GryphonSerialManager;

import java.util.ArrayList;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";
    
    private GryphonSerialManager serialManager;
    private GryphonProtocol gryphonProtocol;
    
    private ScheduledExecutorService pollingExecutor;
    private ArrayList<String> logMessages = new ArrayList<>();
    
    // Handler for delayed operations
    private final android.os.Handler handler = new android.os.Handler();
    
    // Runnable to check for command timeouts
    private final Runnable timeoutChecker = new Runnable() {
        @Override
        public void run() {
            if (gryphonProtocol != null) {
                gryphonProtocol.checkTimeouts();
            }
            // Schedule the next check
            handler.postDelayed(this, 500); // Check every 500ms
        }
    };
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        
        // Initialize the serial manager
        serialManager = GryphonSerialManager.getInstance(this);
        gryphonProtocol = new GryphonProtocol(serialManager);
        
        // Set up the UI components
        setupUI();
        
        // Set up the protocol response listener
        setupResponseListener();
        
        // Start the timeout checker
        handler.post(timeoutChecker);
        
        // Try to connect automatically on startup
        connectToDevice();
    }
    
    private void setupUI() {
        // Connect button
        findViewById(R.id.btnConnect).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                connectToDevice();
            }
        });
        
        // Disconnect button
        findViewById(R.id.btnDisconnect).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                disconnectDevice();
            }
        });
        
        // Poll button
        findViewById(R.id.btnPoll).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                sendPollCommand();
            }
        });
        
        // Reset button
        findViewById(R.id.btnReset).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                sendResetCommand();
            }
        });
        
        // Setup button
        findViewById(R.id.btnSetup).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                sendSetupCommand();
            }
        });
        
        // Coin Type button
        findViewById(R.id.btnCoinType).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                sendCoinTypeCommand();
            }
        });
        
        // Start Polling button
        findViewById(R.id.btnStartPolling).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startPolling();
            }
        });
        
        // Stop Polling button
        findViewById(R.id.btnStopPolling).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                stopPolling();
            }
        });
        
        // Clear Log button
        findViewById(R.id.btnClearLog).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                logMessages.clear();
                updateLogDisplay();
            }
        });
    }
    
    /**
     * Update the UI with the connection status
     */
    private void updateConnectionStatus(String status) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                findViewById(R.id.tvConnectionStatus).post(new Runnable() {
                    @Override
                    public void run() {
                        ((android.widget.TextView) findViewById(R.id.tvConnectionStatus)).setText("Connection Status: " + status);
                    }
                });
            }
        });
    }
    
    /**
     * Update the UI with the last response
     */
    private void updateLastResponse(String response) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                findViewById(R.id.tvLastResponse).post(new Runnable() {
                    @Override
                    public void run() {
                        ((android.widget.TextView) findViewById(R.id.tvLastResponse)).setText("Last Response: " + response);
                    }
                });
            }
        });
    }
    
    /**
     * Add log message and update the log display
     */
    private void addLogMessage(String message) {
        logMessages.add(0, message); // Add to the beginning for newest first
        updateLogDisplay();
    }
    
    /**
     * Update the log display with current messages
     */
    private void updateLogDisplay() {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                StringBuilder logText = new StringBuilder();
                for (String message : logMessages) {
                    logText.append(message).append("\n");
                }
                
                findViewById(R.id.tvLog).post(new Runnable() {
                    @Override
                    public void run() {
                        ((android.widget.TextView) findViewById(R.id.tvLog)).setText(logText.toString());
                    }
                });
            }
        });
    }
    
    /**
     * Set up the response listener for Gryphon protocol commands
     */
    private void setupResponseListener() {
        gryphonProtocol.setResponseListener(new GryphonProtocol.GryphonResponseListener() {
            @Override
            public void onCommandResponse(GryphonProtocol.CommandType command, GryphonProtocol.ResponseStatus status, byte[] responseData) {
                // Update UI with response
                String responseHex = responseData != null ? GryphonSerialManager.bytesToHexString(responseData) : "No data";
                String responseText = command + " - " + status + ": " + responseHex;
                
                updateLastResponse(responseHex);
                addLogMessage(responseText);
                
                // Handle coin acceptance
                if (status == GryphonProtocol.ResponseStatus.COIN_ACCEPTED) {
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Toast.makeText(MainActivity.this, "Coin Accepted!", Toast.LENGTH_SHORT).show();
                        }
                    });
                } else if (status == GryphonProtocol.ResponseStatus.TIMEOUT) {
                    // Handle timeouts with potential reconnection
                    if (!serialManager.isConnected()) {
                        addLogMessage("Connection lost - attempting to reconnect...");
                        reconnectWithDelay(1000); // Try to reconnect after 1 second
                    }
                }
                
                Log.d(TAG, "Command response: " + responseText);
            }
        });
        
        // Set up serial manager listener for connection events
        serialManager.setListener(new GryphonSerialManager.GryphonSerialListener() {
            @Override
            public void onReceiveData(byte[] data) {
                // Data processing is handled by the protocol class
            }
            
            @Override
            public void onConnectionEstablished() {
                updateConnectionStatus("Connected");
                addLogMessage("Connection established with device");
            }
            
            @Override
            public void onConnectionLost() {
                updateConnectionStatus("Disconnected");
                addLogMessage("Connection lost with device");
                stopPolling(); // Stop polling if connection is lost
            }
            
            @Override
            public void onError(Exception e) {
                updateConnectionStatus("Error: " + e.getMessage());
                addLogMessage("Error: " + e.getMessage());
            }
        });
    }
    
    /**
     * Connect to the Gryphon device
     */
    private void connectToDevice() {
        if (serialManager.connectToDevice()) {
            addLogMessage("Connecting to device...");
            
            // Schedule initialization sequence after connection is established
            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    if (serialManager.isConnected()) {
                        initializeConnection();
                    } else {
                        // If not connected yet, try again
                        reconnectWithDelay(1000);
                    }
                }
            }, 1000); // Give the connection a second to establish
        } else {
            addLogMessage("No devices found or permission denied - will retry...");
            reconnectWithDelay(3000); // Try again in 3 seconds
        }
    }
    
    /**
     * Disconnect from the Gryphon device
     */
    private void disconnectDevice() {
        serialManager.closeConnection();
        updateConnectionStatus("Disconnected");
        addLogMessage("Disconnected from device");
        stopPolling(); // Stop polling when disconnected
    }
    
    /**
     * Send a poll command to check for coin acceptance
     */
    private void sendPollCommand() {
        if (serialManager.isConnected()) {
            if (gryphonProtocol.sendPoll()) {
                addLogMessage("Sent Poll command");
            } else {
                addLogMessage("Failed to send Poll command");
            }
        } else {
            addLogMessage("Not connected to device");
        }
    }
    
    /**
     * Send a reset command to the Gryphon
     */
    private void sendResetCommand() {
        if (serialManager.isConnected()) {
            if (gryphonProtocol.sendReset()) {
                addLogMessage("Sent Reset command");
            } else {
                addLogMessage("Failed to send Reset command");
            }
        } else {
            addLogMessage("Not connected to device");
        }
    }
    
    /**
     * Send a setup command to the Gryphon
     */
    private void sendSetupCommand() {
        if (serialManager.isConnected()) {
            if (gryphonProtocol.sendSetup()) {
                addLogMessage("Sent Setup command");
            } else {
                addLogMessage("Failed to send Setup command");
            }
        } else {
            addLogMessage("Not connected to device");
        }
    }
    
    /**
     * Send a coin type command to the Gryphon
     */
    private void sendCoinTypeCommand() {
        if (serialManager.isConnected()) {
            if (gryphonProtocol.sendCoinType()) {
                addLogMessage("Sent Coin Type command");
            } else {
                addLogMessage("Failed to send Coin Type command");
            }
        } else {
            addLogMessage("Not connected to device");
        }
    }
    
    /**
     * Start continuous polling for coins
     */
    private void startPolling() {
        if (!serialManager.isConnected()) {
            addLogMessage("Cannot start polling: Not connected to device");
            return;
        }
        
        // Stop any existing polling job
        stopPolling();
        
        // Start a new polling executor
        pollingExecutor = Executors.newSingleThreadScheduledExecutor();
        pollingExecutor.scheduleAtFixedRate(new Runnable() {
            @Override
            public void run() {
                gryphonProtocol.sendPoll();
            }
        }, 0, 500, TimeUnit.MILLISECONDS); // Poll every 500ms
        
        addLogMessage("Started continuous polling");
    }
    
    /**
     * Stop continuous polling
     */
    private void stopPolling() {
        if (pollingExecutor != null && !pollingExecutor.isShutdown()) {
            pollingExecutor.shutdownNow();
            pollingExecutor = null;
            addLogMessage("Stopped polling");
        }
    }
    
    /**
     * Reconnect to the device after a specified delay
     */
    private void reconnectWithDelay(long delayMillis) {
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                if (!serialManager.isConnected()) {
                    addLogMessage("Attempting reconnection...");
                    connectToDevice();
                }
            }
        }, delayMillis);
    }
    
    /**
     * Initialize the connection by sending the setup command and then 
     * configuring coin types
     */
    private void initializeConnection() {
        if (serialManager.isConnected()) {
            addLogMessage("Initializing device communication...");
            
            // First send a reset command
            if (gryphonProtocol.sendReset()) {
                addLogMessage("Sent Reset command for initialization");
                
                // Wait a bit then send setup 
                handler.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        if (gryphonProtocol.sendSetup()) {
                            addLogMessage("Sent Setup command for initialization");
                            
                            // Wait a bit then configure coin types
                            handler.postDelayed(new Runnable() {
                                @Override
                                public void run() {
                                    if (gryphonProtocol.sendCoinType()) {
                                        addLogMessage("Sent Coin Type command for initialization");
                                        addLogMessage("Device initialization complete");
                                    }
                                }
                            }, 500);
                        }
                    }
                }, 500);
            }
        }
    }
    
    @Override
    protected void onDestroy() {
        super.onDestroy();
        stopPolling();
        handler.removeCallbacks(timeoutChecker); // Stop the timeout checker
        serialManager.destroy(); // Clean up resources
    }
}

File: app/src/main/java/com/example/coingryphon/MainActivity.kt
package com.example.coingryphon

import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.hardware.usb.UsbDevice
import android.hardware.usb.UsbManager
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.coingryphon.serial.GryphonProtocol
import com.example.coingryphon.serial.GryphonSerialManager
import com.example.coingryphon.ui.theme.CoinGryphonTheme
import kotlinx.coroutines.*

class MainActivity : ComponentActivity() {
    private val TAG = "MainActivity"
    
    private lateinit var serialManager: GryphonSerialManager
    private lateinit var gryphonProtocol: GryphonProtocol
    
    // UI state variables
    private var connectionStatus by mutableStateOf("Disconnected")
    private var lastResponse by mutableStateOf("No response yet")
    private var logMessages = mutableStateListOf<String>()
    
    // Coroutine for polling
    private var pollingJob: Job? = null
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        
        // Initialize the serial manager
        serialManager = GryphonSerialManager.getInstance(this)
        gryphonProtocol = GryphonProtocol(serialManager)
        
        // Set up the protocol response listener
        setupResponseListener()
        
        setContent {
            CoinGryphonTheme {
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    GryphonUI()
                }
            }
        }
    }
    
    @Composable
    fun GryphonUI() {
        Scaffold(modifier = Modifier.fillMaxSize()) { innerPadding ->
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(innerPadding)
                    .padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                Text(
                    text = "Coin Gryphon Controller",
                    fontSize = 24.sp,
                    fontWeight = FontWeight.Bold
                )
                
                // Connection status
                Card(
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text(
                            text = "Connection Status: $connectionStatus",
                            fontWeight = FontWeight.Medium
                        )
                        
                        Row(modifier = Modifier.padding(top = 8.dp)) {
                            Button(
                                onClick = { connectToDevice() },
                                modifier = Modifier.padding(end = 8.dp)
                            ) {
                                Text("Connect")
                            }
                            
                            Button(
                                onClick = { disconnectDevice() }
                            ) {
                                Text("Disconnect")
                            }
                        }
                    }
                }
                
                // Control buttons
                Card(
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text(
                            text = "Commands",
                            fontWeight = FontWeight.Medium
                        )
                        
                        Row(modifier = Modifier.padding(top = 8.dp)) {
                            Button(
                                onClick = { sendPollCommand() },
                                modifier = Modifier.padding(end = 8.dp)
                            ) {
                                Text("Poll")
                            }
                            
                            Button(
                                onClick = { sendResetCommand() },
                                modifier = Modifier.padding(end = 8.dp)
                            ) {
                                Text("Reset")
                            }
                            
                            Button(
                                onClick = { sendSetupCommand() },
                                modifier = Modifier.padding(end = 8.dp)
                            ) {
                                Text("Setup")
                            }
                            
                            Button(
                                onClick = { sendCoinTypeCommand() }
                            ) {
                                Text("Coin Type")
                            }
                        }
                        
                        Row(modifier = Modifier.padding(top = 8.dp)) {
                            Button(
                                onClick = { startPolling() },
                                modifier = Modifier.padding(end = 8.dp)
                            ) {
                                Text("Start Polling")
                            }
                            
                            Button(
                                onClick = { stopPolling() }
                            ) {
                                Text("Stop Polling")
                            }
                        }
                    }
                }
                
                // Response display
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .weight(1f)
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text(
                            text = "Last Response: $lastResponse",
                            fontWeight = FontWeight.Medium
                        )
                        
                        Divider(modifier = Modifier.padding(vertical = 8.dp))
                        
                        Text(
                            text = "Log:",
                            fontWeight = FontWeight.Medium
                        )
                        
                        // Log messages in a scrollable list
                        LazyColumn(
                            modifier = Modifier
                                .fillMaxWidth()
                                .weight(1f)
                                .padding(top = 8.dp)
                        ) {
                            items(logMessages.size) { index ->
                                Text(logMessages[logMessages.size - index - 1])
                            }
                        }
                        
                        Button(
                            onClick = { logMessages.clear() },
                            modifier = Modifier.align(Alignment.End)
                        ) {
                            Text("Clear Log")
                        }
                    }
                }
            }
        }
    }
    
    /**
     * Set up the response listener for Gryphon protocol commands
     */
    private fun setupResponseListener() {
        gryphonProtocol.setResponseListener(object : GryphonProtocol.GryphonResponseListener {
            override fun onCommandResponse(
                command: GryphonProtocol.CommandType,
                status: GryphonProtocol.ResponseStatus,
                responseData: ByteArray?
            ) {
                // Update UI with response
                val responseHex = responseData?.let { GryphonSerialManager.bytesToHexString(it) } ?: "No data"
                val responseText = "$command - $status: $responseHex"
                
                lastResponse = responseHex
                logMessages.add(responseText)
                
                // Handle coin acceptance
                if (status == GryphonProtocol.ResponseStatus.COIN_ACCEPTED) {
                    runOnUiThread {
                        Toast.makeText(this@MainActivity, "Coin Accepted!", Toast.LENGTH_SHORT).show()
                    }
                }
                
                Log.d(TAG, "Command response: $responseText")
            }
        })
        
        // Set up serial manager listener for connection events
        serialManager.setListener(object : GryphonSerialManager.GryphonSerialListener {
            override fun onReceiveData(data: ByteArray) {
                // Data processing is handled by the protocol class
            }
            
            override fun onConnectionEstablished() {
                connectionStatus = "Connected"
                logMessages.add("Connection established with device")
            }
            
            override fun onConnectionLost() {
                connectionStatus = "Disconnected"
                logMessages.add("Connection lost with device")
                stopPolling() // Stop polling if connection is lost
            }
            
            override fun onError(e: Exception) {
                connectionStatus = "Error: ${e.message}"
                logMessages.add("Error: ${e.message}")
            }
        })
    }
    
    /**
     * Connect to the Gryphon device
     */
    private fun connectToDevice() {
        if (serialManager.connectToDevice()) {
            logMessages.add("Connecting to device...")
        } else {
            logMessages.add("No USB devices found or permission denied")
        }
    }
    
    /**
     * Disconnect from the Gryphon device
     */
    private fun disconnectDevice() {
        serialManager.closeConnection()
        connectionStatus = "Disconnected"
        logMessages.add("Disconnected from device")
        stopPolling() // Stop polling when disconnected
    }
    
    /**
     * Send a poll command to check for coin acceptance
     */
    private fun sendPollCommand() {
        if (serialManager.isConnected()) {
            if (gryphonProtocol.sendPoll()) {
                logMessages.add("Sent Poll command")
            } else {
                logMessages.add("Failed to send Poll command")
            }
        } else {
            logMessages.add("Not connected to device")
        }
    }
    
    /**
     * Send a reset command to the Gryphon
     */
    private fun sendResetCommand() {
        if (serialManager.isConnected()) {
            if (gryphonProtocol.sendReset()) {
                logMessages.add("Sent Reset command")
            } else {
                logMessages.add("Failed to send Reset command")
            }
        } else {
            logMessages.add("Not connected to device")
        }
    }
    
    /**
     * Send a setup command to the Gryphon
     */
    private fun sendSetupCommand() {
        if (serialManager.isConnected()) {
            if (gryphonProtocol.sendSetup()) {
                logMessages.add("Sent Setup command")
            } else {
                logMessages.add("Failed to send Setup command")
            }
        } else {
            logMessages.add("Not connected to device")
        }
    }
    
    /**
     * Send a coin type command to the Gryphon
     */
    private fun sendCoinTypeCommand() {
        if (serialManager.isConnected()) {
            if (gryphonProtocol.sendCoinType()) {
                logMessages.add("Sent Coin Type command")
            } else {
                logMessages.add("Failed to send Coin Type command")
            }
        } else {
            logMessages.add("Not connected to device")
        }
    }
    
    /**
     * Start continuous polling for coins
     */
    private fun startPolling() {
        if (!serialManager.isConnected()) {
            logMessages.add("Cannot start polling: Not connected to device")
            return
        }
        
        // Stop any existing polling job
        stopPolling()
        
        // Start a new polling coroutine
        pollingJob = CoroutineScope(Dispatchers.IO).launch {
            logMessages.add("Started continuous polling")
            while (isActive) {
                gryphonProtocol.sendPoll()
                delay(500) // Poll every 500ms
            }
        }
    }
    
    /**
     * Stop continuous polling
     */
    private fun stopPolling() {
        pollingJob?.cancel()
        pollingJob = null
        logMessages.add("Stopped polling")
    }
    
    override fun onDestroy() {
        super.onDestroy()
        stopPolling()
        serialManager.destroy() // Clean up resources
    }
}